Provide more crash-info in when checking for runs of zero-bytes

Achierius opened the pull request:
#### 34d00da35e69a4c7f352e7070ab877e7a409ff70
<pre>
Provide more crash-info in when checking for runs of zero-bytes
<a href="https://bugs.webkit.org/show_bug.cgi?id=280547">https://bugs.webkit.org/show_bug.cgi?id=280547</a>
<a href="https://rdar.apple.com/136858657">rdar://136858657</a>

Reviewed by Keith Miller.

Previously, the crash told us that zero-bytes were present in the
destination buffer, but not whether they were present in the source
buffer. This patch fixes that without performance overhead by checking
the source buffer only if a run is detected in the destination buffer.
Moreover, while the crash stack afforded by this diagnostic is more
useful than the &apos;raw&apos; crash would be, it does lose us the ability to see
the actually-corrupted instruction stream: instead of just asserting,
this patch jumps into the JIT buffer to induce a crash.

* Source/JavaScriptCore/jit/ExecutableAllocator.h:
(JSC::dieByJumpingIntoJITBufferWithInfo):
(JSC::performJITMemcpy):

Canonical link: <a href="https://commits.webkit.org/284559@main">https://commits.webkit.org/284559@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/41b06edb267622687006e4697772c879e41b4f25

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69730 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49130 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22483 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73815 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20888 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56931 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20739 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55394 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13856 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72796 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44783 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60136 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35874 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41448 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17576 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19265 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/62846 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63382 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17921 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75539 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/68976 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13964 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17175 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63077 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13999 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60219 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62999 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15500 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11021 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4607 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/90758 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44934 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/16113 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46008 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47279 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45749 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/jit/ExecutableAllocator.h:
```diff
@@ -119,6 +119,71 @@ JS_EXPORT_PRIVATE void dumpJITMemory(const void*, const void*, size_t);
+// while preserving the stack so that we can use that to figure out who might be
+// responsible for generating the corrupted instructions.
+template<typename T, typename U, typename V>
+static NEVER_INLINE NO_RETURN_DUE_TO_CRASH NOT_TAIL_CALLED void dieByJumpingIntoJITBufferWithInfo(int line, void* buffer, size_t offset, size_t size, T info1, U info2, V info3)
```
Nit: You can just do `auto info1, auto info2, auto info3`

Achierius replied:
True! Will fix

kmiller68 commented on Source/JavaScriptCore/jit/ExecutableAllocator.h:
```diff
@@ -119,6 +119,71 @@ JS_EXPORT_PRIVATE void dumpJITMemory(const void*, const void*, size_t);
+#else
+#error "JIT_SCAN_ASSEMBLER_BUFFER_FOR_ZEROES not supported on this platform."
+#endif
+    __builtin_unreachable();
```
Can't we just do 
```
CRASH_WITH_INFO(targetInstr, static_cast<uint64_t>(static_cast<int64_t>(line)), reinterpret_cast<uintptr_t>(buffer), static_cast<uint64_t>(offset | size << 32), wtfCrashArg(info1), wtfCrashArg(info2), wtfCrashArg(info3));
```

Achierius replied:
In that case StabilityTracer would give us the instruction stream of the compiler; we can get a bit more info (i.e. the x86 instructions immediately prior to the ones being corrupted) if we jump into the buffer, essentially imitating what would happen if we didn't catch this corruption during the zero-checking stage. However, since this function is NEVER_INLINE we preserve the prior stack frame and thus get the same stack trace as we were in the zero-checking patch.

Obviously it's a lot of complexity for a bit of diagnostic info, but I couldn't think of any other way to keep that information available.

kmiller68 reviewed (approved):
r=me with comments

Achierius committed:
Provide more crash-info in when checking for runs of zero-bytes
https://bugs.webkit.org/show_bug.cgi?id=280547
rdar://136858657

Reviewed by Keith Miller.

Previously, the crash told us that zero-bytes were present in the
destination buffer, but not whether they were present in the source
buffer. This patch fixes that without performance overhead by checking
the source buffer only if a run is detected in the destination buffer.
Moreover, while the crash stack afforded by this diagnostic is more
useful than the 'raw' crash would be, it does lose us the ability to see
the actually-corrupted instruction stream: instead of just asserting,
this patch jumps into the JIT buffer to induce a crash.

* Source/JavaScriptCore/jit/ExecutableAllocator.h:
(JSC::dieByJumpingIntoJITBufferWithInfo):
(JSC::performJITMemcpy):

Canonical link: https://commits.webkit.org/284559@main
